bin/mail/2016-03-29-ny-cs4all/generate-list-functions.rb

```diff
@@ -0,0 +1,87 @@
+#!/usr/bin/env ruby
+require_relative '../mailing-common/mailing-list-utils'
+require 'active_support'
+require 'active_support/core_ext/object/blank'
+
+# Gather all email addresses *@schools.nyc.gov who are (any of):
+# A. Code Studio teachers who have already been through the K-5 PD.
+# B. Code Studio teachers who have not been through any PD.
+# C. Hour of Code teachers who aren't on Code Studio.
+
+def query_dashboard_teachers(query)
+  {}.tap do |results|
+    DASHBOARD_DB.fetch(query).each do |teacher|
+      email = teacher[:email]
+      results[email] = {email: email, name: teacher[:name], id: teacher[:id]}
+    end
+  end
+end
+
+def query_k5_pd_section_ids
+  DB[:forms].where(kind: 'ProfessionalDevelopmentWorkshop').map(:data).
+    map{|form_data| JSON.parse(form_data)['section_id_s']}.reject(&:blank?).map(&:to_i)
+end
+
+# A. Query Code Studio teachers who have already been through the K-5 PD.
+def query_k5_pd_teachers
+  section_ids = query_k5_pd_section_ids
+  puts "#{section_ids.count} K5 workshop section Ids loaded."
+
+  k5_pd_teachers = query_dashboard_teachers %Q(
+    SELECT DISTINCT culled_users.id, culled_users.email, culled_users.name
+    FROM (
+      SELECT * FROM users WHERE user_type = 'teacher' AND email like '%@schools.nyc.gov'
```

Comment:
For future reference, though [soft] deleted accounts are auto-magically managed by ActiveRecord, [soft] deleted accounts are *not* auto-magically managed by Active Record. Presumably, this and similar queries should have `deleted_at IS NULL` clause as part of the `WHERE`.